pandas/core/groupby/groupby.py

```diff
@@ -691,7 +691,14 @@ def _set_result_index_ordered(self, result):
             result.set_axis(index, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
-        result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+        if (
+            self.dropna
+            and self.axis == 0
+            and len(result.index) < len(self._selected_obj.index)
+        ):
+            result.index = self._selected_obj.index[result.index]
+        else:
+            result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
```

Comment:
this is really hard to understand why you are doing this. 

`.dropna` should modify the `.result_index` on the Grouper object no?

Reply:
you mean the if condition or what we're doing inside the if? 

we have the if because AFAIK `result` has a simple index when `_set_result_index_ordered` is initally called and we need to restore the original index  if it's non-standard or a mulitiindex

If no rows were dropped `set_axis` does the job, but `set_axis` throws in case of shape mismatch between `result.index` and original index 

since `result.index` is a standard index we can get labels of the dropped object by using `result.index` to slice the original index. In the OP it's amounts to doing:
```
In [24]: df.index   # original index                                                                                                                                                                                         
Out[24]: Index(['a', 'b', 'c', 'd'], dtype='object')

In [26]: df.index = df.index[[0, 1, 3]]    # index of dropped object                                                                                                                                                                
    ...: df.index                                                                                                                                                                                                 
Out[29]: Index(['a', 'b', 'd'], dtype='object') 
```
`self.axis` does not propagate correctly rn so checking that does nothing atm